integration/log_integration_test.go

```diff
@@ -0,0 +1,332 @@
+				return fmt.Errorf("queue leaves failed: %s %d", response.Status.Description, response.Status.StatusCode)
+			}
+
+			leaves = leaves[:0] // starting new batch
+		}
+	}
+
+	return nil
+}
+
+func waitForSequencing(treeID int64, client trillian.TrillianLogClient, params testParameters) error {
+	endTime := time.Now().Add(params.sequencingWaitTotal)
+
+	glog.Infof("Waiting for sequencing until: %v", endTime)
+
+	for endTime.After(time.Now()) {
+		req := trillian.GetSequencedLeafCountRequest{LogId: treeID}
+		ctx, cancelFunc := getRPCDeadlineContext()
+		sequencedLeaves, err := client.GetSequencedLeafCount(ctx, &req)
+		cancelFunc()
+
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("Leaf count: %d", sequencedLeaves.LeafCount)
+
+		if sequencedLeaves.LeafCount >= params.leafCount + params.startLeaf {
+			return nil
+		}
+
+		glog.Infof("Leaves sequenced: %d. Still waiting ...", sequencedLeaves.LeafCount)
+
+		time.Sleep(params.sequencingPollWait)
+	}
+
+	return errors.New("wait time expired")
+}
+
+func readbackLogEntries(logID int64, client trillian.TrillianLogClient, params testParameters) (map[int64]*trillian.LogLeaf, error) {
+	currentLeaf := int64(0)
+	leafMap := make(map[int64]*trillian.LogLeaf)
+
+	// Build a map of all the leaf data we expect to have seen when we've read all the leaves.
+	// Have to work with strings because slices can't be map keys. Sigh.
+	leafDataPresenceMap := make(map[string]bool)
+
+	for l := int64(0); l < params.leafCount; l++ {
+		leafDataPresenceMap[fmt.Sprintf("Leaf %d", l + params.startLeaf)] = true
+	}
+
+	for currentLeaf < params.leafCount {
+		hasher := merkle.NewRFC6962TreeHasher(crypto.NewSHA256())
+
+		// We have to allow for the last batch potentially being a short one
+		numLeaves := params.leafCount - currentLeaf
+
+		if numLeaves > int64(params.readBatchSize) {
+			numLeaves = int64(params.readBatchSize)
+		}
+
+		glog.Infof("Reading %d leaves from %d ...", numLeaves, currentLeaf + params.startLeaf)
+		req := makeGetLeavesByIndexRequest(logID, currentLeaf + params.startLeaf, numLeaves)
+		ctx, cancelFunc := getRPCDeadlineContext()
+		response, err := client.GetLeavesByIndex(ctx, req)
+		cancelFunc()
+
+		if err != nil {
+			return nil, err
+		}
+
+		if response.Status == nil || response.Status.StatusCode != trillian.TrillianApiStatusCode_OK {
+			return nil, fmt.Errorf("read leaves failed: %s %d", response.Status.Description, response.Status.StatusCode)
+		}
+
+		// Check we got the right leaf count
+		if int64(len(response.Leaves)) == 0 {
+			return nil, fmt.Errorf("expected %d leaves log returned none", numLeaves)
+		}
+
+		// Check the leaf contents make sense. Can't rely on exact ordering as queue timestamps will be
+		// close between batches and identical within batches.
+		for l := 0; l < len(response.Leaves); l++ {
+			// Check for duplicate leaf index in response data - should not happen
+			if _, ok := leafMap[response.Leaves[l].LeafIndex]; ok {
```

Comment:
`response.Leaves[l].LeafIndex` is used quite a lot in this loop, could assign it to a variable up here to reduce code, or better, use a `range()` style loop to do it for you.